### PR TITLE
CDAP-20179 Appfabric and Preview pods to watch only the resources they create

### PR DESCRIPTION
**Pull request by ameya0111**
JIRA - https://cdap.atlassian.net/browse/CDAP-20179

    - Adding a parent pod label to the k8s resources created by AppFabric and Preview pods.
    - Modifying selector so that pod watches only the resources with parent pod label equal to the pod name.
    - Throws an exception, if the pod is not part of stateful set.

Testing - 

With the code changes the labels appear in the sts as follows - 

```
Labels:        cdap.container=PreviewRunnerTwillRunnable
                    cdap.instance=ameya-cdap-20179-debug
                    cdap.twill.app=preview-runner
                    cdap.twill.parent.pod=cdap-ameya-cdap-20179-debug-preview-0
                    cdap.twill.run.id=3ae76a5b-f107-443e-8a0b-227bd18b0d5b
                    cdap.twill.runner=k8s
                    k8s.namespace=default
```
                    **cdap.twill.parent.pod=cdap-ameya-cdap-20179-debug-preview-0**

Both the preview-runner sts and the task-worker sts were deleted and logs observed in appFabric and preview pods - 

```
➜  cdap git:(bugfix/CDAP-20179) ✗ kb delete sts cdap-ameya-cdap-20179-debug-preview-runne-2e81d3027d
statefulset.apps "cdap-ameya-cdap-20179-debug-preview-runne-2e81d3027d" deleted
➜  cdap git:(bugfix/CDAP-20179) ✗ kb delete sts cdap-ameya-cdap-20179-debug-task-worker-3-0678b67462
statefulset.apps "cdap-ameya-cdap-20179-debug-task-worker-3-0678b67462" deleted
```

AppFabric Logs - 

```
➜  ~ kb logs cdap-ameya-cdap-20179-debug-appfabric-0 appfabric -f | grep "Deleting StatefulSet"
2023-01-12 03:32:37,028 - DEBUG [kube-statefulsets-watch:i.c.c.k.r.KubeTwillController@485] - Deleting StatefulSet cdap-ameya-cdap-20179-debug-task-worker-3-0678b67462
2023-01-12 03:32:37,045 - DEBUG [kube-statefulsets-watch:i.c.c.k.r.KubeTwillController@485] - Deleting StatefulSet cdap-ameya-cdap-20179-debug-task-worker-3-0678b67462
^C


```

Preview pod Logs - 

```
➜  cdap git:(bugfix/CDAP-20179) ✗  kb logs cdap-ameya-cdap-20179-debug-preview-0 preview -f | grep "Deleting StatefulSet"
2023-01-12 03:32:07,322 - DEBUG [kube-statefulsets-watch:i.c.c.k.r.KubeTwillController@485] - Deleting StatefulSet cdap-ameya-cdap-20179-debug-preview-runne-2e81d3027d
2023-01-12 03:32:07,344 - DEBUG [kube-statefulsets-watch:i.c.c.k.r.KubeTwillController@485] - Deleting StatefulSet cdap-ameya-cdap-20179-debug-preview-runne-2e81d3027d
^C

```

### CODE REVIEW

**Review by masoud-io — COMMENTED** [review]
This approach is a bit hacky and brittle. We should aim for a more generic approach.

**Comment by ameya0111** [comment]
> This approach is a bit hacky and brittle. We should aim for a more generic approach.

I am thinking something on the lines of - 

Each pod is configured with a list of resources it watches. 
But trying to come up with a way , so that these lists are disjoint.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java [hunk 1]
```diff
@@ -194,6 +197,19 @@ public KubeTwillRunnerService(MasterEnvironmentContext masterEnvContext, ApiClie
     this.workloadIdentityProvider = workloadIdentityProvider;
   }
 
+  private String generateSelector() {
```
Comment (tivv) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: Do you try to have two different preview sets for manager and appfabric? If so, why? Why does app fabric need to manage this stateful set

Comment (ameya0111) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: No, the plan is for AppFabric to not manage it. Only Preview manager should.

**Comment by chtyim** [comment]
What is the problem with multiple pods trying to delete the resources? Is it an optimization or is it causing some wrong logic to happen? Deletion is idempotent.

**Comment by tivv** [comment]
More details in JIRA, but two managers start to issue crreation / deletion calls according to their pipelines with high fequency (E.g. set created and then 4 deletion requests issued within 500ms).

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java (hunk 1 above)
Comment (tivv) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: Then why selector needs to know the current pod and such?

Comment (chtyim) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: Can we have only app-fabric to manage all?

Comment (ameya0111) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: I will refactor this PR to better provide the selector with information as to what to watch out for. So that the information is coming from the Pods. Used this PR to get comments to confirm functionality.

Will remove the hard-coded values from KubeTwillRunnerService.

Comment (ameya0111) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: > Can we have only app-fabric to manage all?

The preview-runners are created by the preview pod. So the separation seems cleaner if we let preview manager pod watch these resources. And app fabric watches the resources it created.

I am modifying this PR, so that the hacky approach in this first PR is replaced.

**Comment by ameya0111** [comment]
> This approach is a bit hacky and brittle. We should aim for a more generic approach.

I have modified this PR to have a more generic approach. Please let me know your thoughts in the comments.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java (hunk 1 above)
Comment (ameya0111) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: > Then why selector needs to know the current pod and such?

I have modified the commit so that the approach is generic. 
In this PR, there is a parent.pod label. 

As we discussed offline, it would be worth it to explore if we can have a service name instead of pod name. So that if we try for replication of app-fabric or preview pods in the future, it would be good. 

I am looking at possible ways to do so.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java [hunk 2]
```diff
@@ -194,6 +197,27 @@ public KubeTwillRunnerService(MasterEnvironmentContext masterEnvContext, ApiClie
     this.workloadIdentityProvider = workloadIdentityProvider;
   }
 
+  /**
+   * Returns true if the pod is owned by a stateful set.
+   */
+  private boolean isStatefulSetPod() {
+    return this.podInfo.getOwnerReferences().size() == 1
```
Comment (masoud-io) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: can owner reference be >1?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java [hunk 3]
```diff
@@ -194,6 +197,27 @@ public KubeTwillRunnerService(MasterEnvironmentContext masterEnvContext, ApiClie
...
+      && this.podInfo.getOwnerReferences().get(0).getKind().equals("StatefulSet");
+  }
+
+  /**
+   * Generates k8s label selector which selects all runs started by the k8s twill runner that has the run id label.
+   * If the pod comes from a statefulset it also adds a PARENT_POD_LABEL to the selector.
+   */
+  private String generateSelector() {
+    if (isStatefulSetPod()) {
```
Comment (masoud-io) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: Why do we need to check stateful set only?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java [hunk 4]
```diff
@@ -218,6 +242,9 @@ public TwillPreparer prepare(TwillApplication application) {
     labels.put(RUNNER_LABEL, RUNNER_LABEL_VAL);
     labels.put(APP_LABEL, spec.getName());
     labels.put(RUN_ID_LABEL, runId.getId());
+    if (isStatefulSetPod()) {
```
Comment (masoud-io) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: why not be consistent and put parent pod label for any resource?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java [hunk 5]
```diff
@@ -125,6 +126,7 @@ public class KubeTwillRunnerService implements TwillRunnerService, NamespaceList
   private static final String CDAP_NAMESPACE_LABEL = "cdap.namespace";
   private static final String NAMESPACE_CPU_LIMIT_PROPERTY = "k8s.namespace.cpu.limits";
   private static final String NAMESPACE_MEMORY_LIMIT_PROPERTY = "k8s.namespace.memory.limits";
+  public static final String PARENT_POD_LABEL = "cdap.twill.parent.pod";
```
Comment (masoud-io) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: nit: `cdap. ` is not needed

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java (hunk 4 above)
Comment (ameya0111) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: Note - we are adding this label for any resource. The check is being done on whether the Current pod that is creating the resources is part of a stateful set - i.e AppFabric or Preview.

Reason we are not doing this for resources created by pods coming from Deployments is-  

In case of deployment resources - the names may not remain constant. 
In stateful resources - once a replica goes down, it comes up again with the same name. 

> Unlike a Deployment, a StatefulSet maintains a sticky identity for each of their Pods. 
(https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)

Based on the above, we thought of adding this parent pod label only if the parent pod is part of stateful set.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java (hunk 2 above)
Comment (ameya0111) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: Yes, according to this article - https://stackoverflow.com/questions/59759438/when-will-a-kube-object-have-more-than-one-ownerreferences , it is a possibility. But it applies for a very general case.
The owner reference in case of AppFabric pod and Preview pods will point to the single stateful set that created them.

Since it is a list, I am adding this check before doing a get(0).

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java (hunk 3 above)
Comment (ameya0111) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: As mentioned in below comment, we do not want this functionality to exist in case we are doing this in Deployment sets, because the Pod name will not be sticky identity.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java (hunk 2 above)
Comment (masoud-io) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: Agree, but calling isStatefulSetPod in future on a sts pod with multiple owner reference will return false which is wrong. Why not checking >=1 ?

**Review by masoud-io — APPROVED** [review]
One comment; LGTM otherwise.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java (hunk 2 above)
Comment (ameya0111) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: I have modified this function to - 



`    return this.podInfo.getOwnerReferences().stream().anyMatch(r -> r.getKind().equals("StatefulSet"));
`

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java [hunk 6]
```diff
@@ -194,6 +197,31 @@ public KubeTwillRunnerService(MasterEnvironmentContext masterEnvContext, ApiClie
...
+   * (https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)
+   * Therefore, even if the pod goes down, it will come up with the same pod name and this selector can
+   * be continued to be used.
+   */
+  private String generateSelector() {
+    if (isPodControlledByStatefulSet()) {
+      return String.format("%s=%s,%s=%s,%s", PARENT_POD_LABEL, podInfo.getName(),
+                  RUNNER_LABEL, RUNNER_LABEL_VAL, RUN_ID_LABEL);
+    } else {
+      return String.format("%s=%s,%s", RUNNER_LABEL, RUNNER_LABEL_VAL, RUN_ID_LABEL);
```
Comment (tivv) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: I don't think we need to allow it at all, just error out

Comment (ameya0111) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java]: Done.

**Review by tivv — APPROVED** [review]

**Comment by ameya0111** [comment]
@masoud-io 
One behavior that needs to be checked for, before this is cherry-picked. 

In some environments such as a Cloud Data Fusion instance, 
when a 
CDAP Upgrade operation is applied - from older CDAP to 6.8.1,
The existing Task Worker, System Worker, Preview pods will not have the new parent.pod labels. 

These need to be cleaned up during the upgrades. This edge case came to my mind yesterday, and I was not able to test this fully.

 Please test if during the CDAP upgrade those pods are cleaned up.

If in current behavior they are cleaned up, then it is fine.
Else there might be orphaned pods in the system, which are not watched by AppFabric or Preview Pod.
